### PR TITLE
feat: support LB priority for non xRoute endpoints 

### PR DESCRIPTION
**Pull request by alexwo**
### What does this PR do?
It enables users to configure failover backends.

### How is this achieved?
The proposal involves adding a Failover boolean property to backend refs, what allows to indicate whether the backend is designated as a failover.

This approach is illustrated in this PR for external extProc.

As per issue #3055

### CODE REVIEW

File: api/v1alpha1/wasm_types.go [hunk 1]
```diff
@@ -49,9 +49,9 @@ type Wasm struct {
 	// +kubebuilder:default=false
 	FailOpen *bool `json:"failOpen,omitempty"`
 
-	// Priority defines the location of the Wasm extension in the HTTP filter chain.
```
Comment (guydc) [inline api/v1alpha1/wasm_types.go:52]: I think that this should not change in this PR. This priority is related to location in the filter chain, not backend failover.

File: api/v1alpha1/shared_types.go [hunk 2]
```diff
@@ -473,6 +473,9 @@ type BackendRef struct {
 	// BackendObjectReference references a Kubernetes object that represents the backend.
 	// Only Service kind is supported for now.
 	gwapiv1.BackendObjectReference `json:",inline"`
+	// Failover This indicates whether the backend is designated as a failover.
+	// Multiple failover backends can be configured for a single BackendService.⬤
```
Comment (guydc) [inline api/v1alpha1/shared_types.go:477]: nit: the ⬤ character is redundant

**Review by guydc — COMMENTED** [review]
Can we also add an XDS translation test?

**Comment by alexwo** [comment]
> Can we also add an XDS translation test?

added

**Comment by alexwo** [comment]
> Can we also add an XDS translation test?

added

**Comment by guydc** [comment]
overall LGTM. 

In terms of naming, I think that failover is a common industry term used by Akamai, Cloudflare, GCP, AWS and Azure in this context. Nginx and HAProxy use "backup". 

I think that "backup" can be a bit vague in this context:
- It's associated with data backup/restore operations
- It doesn't express the fact that the "backup" backend will automatically handle traffic in case of primary backend failure. Users may anticipate that they need to "enable" a backup server as a primary in case of failure.

**Review by guydc — DISMISSED** [review]

File: api/v1alpha1/shared_types.go [hunk 3]
```diff
@@ -473,6 +473,9 @@ type BackendRef struct {
 	// BackendObjectReference references a Kubernetes object that represents the backend.
 	// Only Service kind is supported for now.
 	gwapiv1.BackendObjectReference `json:",inline"`
+	// Failover This indicates whether the backend is designated as a failover.
+	// Multiple failover backends can be configured for a single BackendService.
```
Comment (zirain) [inline api/v1alpha1/shared_types.go:477]:
```suggestion
	// Multiple failover backends can be configured for a single BackendService.
	// +optional
```

File: api/v1alpha1/shared_types.go [hunk 4]
```diff
@@ -473,6 +473,10 @@ type BackendRef struct {
...
```
Comment (arkodg) [inline api/v1alpha1/shared_types.go:477]:
```suggestion
	// Multiple failover backends can be configured.
```

File: api/v1alpha1/shared_types.go [hunk 5]
```diff
@@ -473,6 +473,10 @@ type BackendRef struct {
...
+	// +optional
```
Comment (arkodg) [inline api/v1alpha1/shared_types.go:478]: "Its strongly recommended to configure active or passive healthchecks to ensure failover can be detected when the active backends are unhealthy and can be readjusted once the primary backends are healthy again. The overprovisioning factor is set to 1.4 i.e. the failover backends will receive traffic only once the health of the active backends drops below 72%"

File: internal/gatewayapi/securitypolicy.go [hunk 6]
```diff
@@ -851,7 +854,7 @@ func (t *Translator) buildExtAuth(policy *egv1a1.SecurityPolicy, resources *Reso
 	authority = backendRefAuthority(resources, backendRef, policy)
 	pnn := utils.NamespacedName(policy)
 	if ds, err = t.processExtServiceDestination(
-		backendRef,
+		&egv1a1.BackendRef{BackendObjectReference: *backendRef, Failover: priority},
```
Comment (arkodg) [inline internal/gatewayapi/securitypolicy.go:857]: can we modify this to `Failover: ptr.Deref(grpc.BackendRefs[0].Failover , false)`

Comment (arkodg) [reply internal/gatewayapi/securitypolicy.go:857]: oh wait, it looks like this line is typecasting from bool from int32 ?

Comment (alexwo) [reply internal/gatewayapi/securitypolicy.go:857]: We don't have to type cast (its bool type), priority should be named failover.

File: api/v1alpha1/shared_types.go (hunk 4 above)
Comment (alexwo) [reply api/v1alpha1/shared_types.go]: updated

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**